File: detekt/wrapper/src/main/java/io/buildfoundation/bazel/detekt/execute/WorkerExecutable.java
package io.buildfoundation.bazel.detekt.execute;

import io.buildfoundation.bazel.detekt.value.WorkRequest;
import io.buildfoundation.bazel.detekt.value.WorkResponse;

public interface WorkerExecutable {

    WorkResponse execute(WorkRequest request);

    final class Impl implements WorkerExecutable {

        private final Executable executable;

        public Impl(Executable executable) {
            this.executable = executable;
        }

        @Override
        public WorkResponse execute(WorkRequest request) {
            String[] arguments = new String[request.arguments.size()];
            request.arguments.toArray(arguments);

            ExecutableResult result = executable.execute(arguments);

            WorkResponse response = new WorkResponse();
            response.requestId = request.requestId;
            response.output = result.output();
            response.exitCode = result.statusCode();

            return response;
        }
    }
}

File: detekt/wrapper/src/main/kotlin/bazel/detekt/WorkerExecutable.kt
package bazel.detekt

import bazel.worker.WorkerProtocol.WorkRequest
import bazel.worker.WorkerProtocol.WorkResponse

internal interface WorkerExecutable {

    fun execute(request: WorkRequest): WorkResponse

    class Impl(private val executable: Executable) : WorkerExecutable {

        override fun execute(request: WorkRequest): WorkResponse {
            val result = executable.execute(request.argumentsList.toTypedArray())

            val resultOutput = when (result) {
                is Result.Success -> WorkResponse.getDefaultInstance().output
                is Result.Failure -> result.description
            }

            return WorkResponse.newBuilder()
                    .setRequestId(request.requestId)
                    .setOutput(resultOutput)
                    .setExitCode(result.consoleStatusCode)
                    .build()
        }
    }
}